File: src/main/java/com/example/chatService/entity/ChatRoom.java
package com.example.chatService.entity;

import com.example.chatService.dto.RoomType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "chat_rooms")
@Getter @Setter
@NoArgsConstructor
public class ChatRoom {

    @Id
    @Column(length = 36)
    private String roomId; // UUID

    @Column(nullable = false)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RoomType type;

    @Column(nullable = false)
    private int currentCount;

    @Column(nullable = false)
    private int maxParticipants;

    @Version
    private long version;

    @Column(nullable = false)
    private boolean largeRoom;

    @Column(nullable = false)
    private String ownerUserId;

    @Column(nullable = false)
    private OffsetDateTime createdAt;

    public static ChatRoom create(
            String name,
            RoomType type,
            int maxParticipants,
            String ownerUserId
    ) {
        ChatRoom r = new ChatRoom();
        r.roomId = UUID.randomUUID().toString();
        r.name = name;
        r.type = type;
        r.maxParticipants = maxParticipants;
        r.largeRoom = maxParticipants >= 100;
        r.ownerUserId = ownerUserId;
        r.createdAt = OffsetDateTime.now();
        return r;
    }

    public void increaseCount() {
        if (currentCount >= maxParticipants) {
            throw new IllegalStateException("Room is full");
        }
        this.currentCount++;
    };

    public void decreaseCount() {
//        if (currentCount <= 0) {
//            throw new IllegalStateException("Room count underflow");
//        }
        this.currentCount--;
    };
}

File: src/main/java/com/example/chatService/entity/ChatRoom.kt
package com.example.chatService.entity

import com.example.chatService.dto.RoomType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.Id
import jakarta.persistence.Table
import jakarta.persistence.Version
import java.time.OffsetDateTime
import java.util.UUID

@Entity
@Table(name = "chat_rooms")
class ChatRoom(

        @Id
        @Column(length = 36)
        var roomId: String = "",

        @Column(nullable = false)
        var name: String = "",

        @Enumerated(EnumType.STRING)
        @Column(nullable = false)
        var type: RoomType? = null,

        @Column(nullable = false)
        var currentCount: Int = 0,

        @Column(nullable = false)
        var maxParticipants: Int = 0,

        @Version
        var version: Long = 0L,

        @Column(nullable = false)
        var largeRoom: Boolean = false,

        @Column(nullable = false)
        var ownerUserId: String = "",

        @Column(nullable = false)
        var createdAt: OffsetDateTime = OffsetDateTime.now()
) {

    fun increaseCount() {
        if (currentCount >= maxParticipants) {
            throw IllegalStateException("Room is full")
        }
        currentCount++
    }

    fun decreaseCount() {
        currentCount--
    }

    companion object {
        fun create(
                name: String,
                type: RoomType,
                maxParticipants: Int,
                ownerUserId: String
        ): ChatRoom {
            return ChatRoom().apply {
                roomId = UUID.randomUUID().toString()
                this.name = name
                this.type = type
                this.maxParticipants = maxParticipants
                largeRoom = maxParticipants >= 100
                this.ownerUserId = ownerUserId
                createdAt = OffsetDateTime.now()
            }
        }
    }
}